code/utils/RandomRange.h

```diff
@@ -261,4 +346,185 @@ UniformRange<Value> parseUniformRange(Value min = std::numeric_limits<Value>::mi
 		return UniformRange<Value>(valueList[0], valueList[1]);
 	}
 }
+
+class CurveNumberDistribution {
+  public:
+	using result_type = float;
+	using param_type = struct {
+		int curve;
+		float min;
+		float max;
+	};
+	param_type m_param;
+	inline CurveNumberDistribution() : CurveNumberDistribution(param_type{-1, 0.f, 1.f}) {}
+	inline CurveNumberDistribution(param_type curve) : m_param(curve) {}
+	inline void reset() {}
+	inline param_type param()
+	{
+		return m_param;
+	}
+	inline void param(param_type p)
+	{
+		m_param = p;
+	}
+	template <typename Generator>
+	inline result_type operator()(Generator& generator, const param_type& param)
+	{
+		if (param.curve < 0) {
+			return 0.f;
+		}
+		float max_integral = Curves[param.curve].GetValueIntegrated(Curves[param.curve].keyframes.back().pos.x);
+		float rand =
+			std::generate_canonical<float, std::numeric_limits<float>::digits, Generator>(generator) * max_integral;
+		float lower_bound = Curves[param.curve].keyframes.front().pos.x;
+		float upper_bound = Curves[param.curve].keyframes.back().pos.x;
+		float curve_min = lower_bound;
+		float curve_max = upper_bound;
+		for (size_t count = 0; count < 16; count++) {
+			float current_pos = (lower_bound + upper_bound) / 2.f;
+			float current_value = Curves[param.curve].GetValueIntegrated(current_pos);
+			if (fl_equal(current_value, rand, max_integral * 0.01f)) {
+				// remap the values to the distribution's range
+				return (current_pos - curve_min) / (curve_max - curve_min) * (param.max - param.min) + param.min;
+			}
+			if (current_value > rand) {
+				upper_bound = current_pos;
+			} else {
+				lower_bound = current_pos;
+			}
+		}
+		// remap the values to the distribution's range
+		return (((lower_bound + upper_bound) / 2.f) - curve_min) / (curve_max - curve_min) * (param.max - param.min) + param.min;
```

Comment:
There's a possible divide by 0 here if the curve's y values are 0.

Reply:
I think I'll insert some checks to just reject any curves that go below 0 being used for distributions during parsing. Same with curves with zero height.